### PR TITLE
Add remote info to the HLRC

### PR DESCRIPTION
**Pull request by AntonShuvaev**
The remote info call is a rest only API call, but should still be added
to the rest client. This PR adds it as well as relevant tests.

Ref #47678

### CODE REVIEW

**Comment by hub-cap** [comment]
great work so far. Lets go one step further with it tho and copy out the request/response from server, and put it into the HLRC's module, so that we dont have to use/modify the server side code. This will all be generated one day so we want to keep any changes away from server. While it might be a bit tedious now, its something our team is working on doing (code generation, eventually).

This will require you writing some new tests (you can look at some of the other PRs that have been recently merged) for the request and response. The request is generally easy, and teh response is more work. Below are the tests you can inherit from and use as examples.

https://github.com/elastic/elasticsearch/blob/master/client/rest-high-level/src/test/java/org/elasticsearch/client/AbstractResponseTestCase.java
https://github.com/elastic/elasticsearch/blob/master/client/rest-high-level/src/test/java/org/elasticsearch/client/AbstractRequestTestCase.java

**Comment by martijnvg** [comment]
@j-bean Thanks for adding this api to the HLRC! Can you copy the `RemoteInfoRequest` and `RemoteInfoResponse` to the hlrc module for the reasons @hub-cap mentioned? You can also remove any thing from these copies that the hlrc doesn't need (for example extending from `ActionRequest` and `ActionResponse`). Besides this this PR looks great!

**Comment by j-bean** [comment]
@martijnvg Sure, I'll fix all comments ASAP.

**Comment by j-bean** [comment]
@martijnvg I fixed all issues and updated the PR.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/CCRIT.java [hunk 1]
```diff
@@ -74,27 +69,7 @@
...
-        assertThat(updateSettingsResponse.isAcknowledged(), is(true));
-
-        assertBusy(() -> {
-            Map<?, ?> localConnection = (Map<?, ?>) toMap(client()
-                .performRequest(new Request("GET", "/_remote/info")))
-                .get("local_cluster");
-            assertThat(localConnection, notNullValue());
-            assertThat(localConnection.get("connected"), is(true));
-        });
+        setupRemoteClusterConfig("local_cluster");
```
Comment (martijnvg) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/CCRIT.java]: 👍 thanks for cleaning this up.

File: server/src/main/java/org/elasticsearch/transport/RemoteConnectionInfo.java [hunk 2]
```diff
@@ -20,30 +20,74 @@
...
 
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.optionalConstructorArg;
+
 /**
  * This class encapsulates all remote cluster information to be rendered on
  * {@code _remote/info} requests.
  */
 public final class RemoteConnectionInfo implements ToXContentFragment, Writeable {
+    private static final String CONNECTED = "connected";
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/transport/RemoteConnectionInfo.java]: I think that `RemoteConnectionInfo`, `ModeInfo` and subclasses should be copied to hlrc module too for the same reason why the response class was copied to hlrc module.

**Review by martijnvg — COMMENTED** [review]
Thanks @j-bean. I left another comment.

**Comment by j-bean** [comment]
Made changes as per review comments.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/remote/RemoteInfoResponse.java [hunk 3]
```diff
@@ -39,7 +39,7 @@
         infos = in.readList(RemoteConnectionInfo::new);
     }
 
-    RemoteInfoResponse(Collection<RemoteConnectionInfo> infos) {
+    public RemoteInfoResponse(Collection<RemoteConnectionInfo> infos) {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/remote/RemoteInfoResponse.java]: I think all changes to the server gradle module can be undone now?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/cluster/RemoteInfoResponseTests.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
+package org.elasticsearch.client.cluster;
```
Comment (martijnvg) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/cluster/RemoteInfoResponseTests.java]: this file is missing apache2 copyright header.

**Review by martijnvg — COMMENTED** [review]
Thanks for updating this pr! I left 2 small comments.

**Comment by martijnvg** [comment]
@elasticmachine please test this

**Comment by martijnvg** [comment]
@elasticmachine test this please

File: server/src/main/java/org/elasticsearch/action/admin/cluster/remote/RemoteInfoResponse.java (hunk 3 above)
Comment (j-bean) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/remote/RemoteInfoResponse.java]: It is used in `RemoteInfoResponseTests`.

Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/remote/RemoteInfoResponse.java]: Ah! I missed that.

**Comment by martijnvg** [comment]
@elasticmachine test this please

**Comment by martijnvg** [comment]
@j-bean Thank you for your contribution. I will backport this now to the 7.x branch.

**Comment by martijnvg** [comment]
@j-bean I had to revert this PR due to [a test failure](https://gradle-enterprise.elastic.co/s/6uhgi5okby7ko/tests/p7wcnrqviliea-lmdccaowz4oiu). I will open a new pr that adds re-adds this API.

**Pull request merged**